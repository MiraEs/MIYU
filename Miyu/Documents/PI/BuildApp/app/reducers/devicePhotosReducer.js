'use strict';
import { handleActions } from 'redux-actions';
import {
	ADD_PHOTO,
	FILTER_PHOTOS_SELECTED,
	LOAD_DEVICE_PHOTOS_FAIL,
	LOAD_DEVICE_PHOTOS_SUCCESS,
	REPLACE_SELECTED_PHOTOS,
	TOGGLE_FETCHING_PHOTOS,
	TOGGLE_PHOTO_SELECTED,
} from '../constants/constants';

const initialState = {
	cursor: '',
	error: '',
	hasNextPage: false,
	isFetchingNextPage: false,
	photos: [],
	selectedPhotos: [],
};

const devicePhotosReducer = handleActions({
	[ADD_PHOTO]: (state, { payload: uri }) => {
		const photos = [{
			image: {
				uri,
			},
			isSelected: true,
		}, ...state.photos];
		const selectedPhotos = photos.filter((photo) => {
			return photo.isSelected;
		});
		return {
			...state,
			selectedPhotos,
			photos,
		};
	},
	[FILTER_PHOTOS_SELECTED]: (state, { payload }) => {
		return {
			...state,
			data: payload.filter((photo) => photo.selected),
		};
	},
	[LOAD_DEVICE_PHOTOS_FAIL]: (state, { payload: error }) => {
		return {
			...state,
			error,
		};
	},
	[LOAD_DEVICE_PHOTOS_SUCCESS]: (state, { payload }) => {
		const photos = payload.edges.map((photo) => {
			const selectedMatch = state.selectedPhotos.find((selectedPhoto) => {
				return selectedPhoto.image.uri === photo.node.image.uri;
			});
			return {
				...photo.node,
				isSelected: selectedMatch !== undefined,
			};
		});
		return {
			...state,
			photos: payload.isFirstPage ? photos : [...state.photos, ...photos],
			hasNextPage: payload.page_info.has_next_page,
			cursor: payload.page_info.has_next_page ? payload.page_info.end_cursor : initialState.cursor,
			isFetchingNextPage: false,
		};
	},
	[REPLACE_SELECTED_PHOTOS]: (state, { payload: selectedPhotos }) => {
		return {
			...state,
			selectedPhotos,
		};
	},
	[TOGGLE_FETCHING_PHOTOS]: (state) => {
		return {
			...state,
			isFetchingNextPage: !state.isFetchingNextPage,
		};
	},
	[TOGGLE_PHOTO_SELECTED]: (state, { payload: uri }) => {
		const selectedPhotos = [];
		const photos = state.photos.map((photo) => {
			const isSelected = photo.image.uri === uri ? !photo.isSelected : photo.isSelected;
			if (isSelected) {
				selectedPhotos.push(photo.image.uri);
			}
			return {
				...photo,
				isSelected,
			};
		});

		return {
			...state,
			photos,
			selectedPhotos,
		};
	},
}, initialState);

export default devicePhotosReducer;
