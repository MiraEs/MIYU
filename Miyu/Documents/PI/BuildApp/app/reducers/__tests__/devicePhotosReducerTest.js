import devicePhotosReducer from '../devicePhotosReducer';

const photo1 = {
	image: {
		uri: '/photo/uri',
	},
	isSelected: true,
};

const photo2 = {
	image: {
		uri: '/other/photo/uri',
	},
	isSelected: false,
};

const photo3 = {
	image: {
		uri: '/yet/another/photo/uri',
	},
	isSelected: true,
};

const initialState = {
	cursor: '',
	error: '',
	hasNextPage: false,
	isFetchingNextPage: false,
	photos: [],
	selectedPhotos: [],
};

describe('devicePhotosReducer', () => {
	it('should handle ADD_PHOTO', () => {
		const action = {
			type: 'ADD_PHOTO',
			payload: '/test/uri',
		};
		const result = devicePhotosReducer(initialState, action);
		expect(result).toMatchSnapshot();
	});
	it('should handle FILTER_SELECTED_PHOTOS', () => {
		const action = {
			type: 'FILTER_PHOTOS_SELECTED',
			payload: [{
				uri: '/selected/photo',
				selected: true,
			}, {
				uri: '/not/selected/photo',
				selected: false,
			}],
		};
		const result = devicePhotosReducer(initialState, action);
		expect(result).toMatchSnapshot();
	});
	it('should handle LOAD_DEVICE_PHOTOS_FAIL', () => {
		const error = new Error('load device photos test error');
		const action = {
			type: 'LOAD_DEVICE_PHOTOS_FAIL',
			payload: error,
		};
		const result = devicePhotosReducer(initialState, action);
		expect(result).toMatchSnapshot();
	});
	it('should handle LOAD_DEVICE_PHOTOS_SUCCESS', () => {
		const action = {
			type: 'LOAD_DEVICE_PHOTOS_SUCCESS',
			payload: {
				edges: [{
					node: {
						image: {
							uri: '/uri/1',
						},
					},
				}],
				page_info: {
					has_next_page: true,
					end_cursor: 'u-u-i-d',
				},
			},
		};
		const result = devicePhotosReducer(initialState, action);
		expect(result).toMatchSnapshot();
	});
	it('should handle LOAD_DEVICE_PHOTOS_SUCCESS with selectedPhotos and one page', () => {
		const action = {
			type: 'LOAD_DEVICE_PHOTOS_SUCCESS',
			payload: {
				edges: [{
					node: photo1,
				}],
				page_info: {
					has_next_page: false,
					end_cursor: 'u-u-i-d',
				},
				isFirstPage: true,
			},
		};
		const initial = {
			...initialState,
			selectedPhotos:[photo1],
		};
		const result = devicePhotosReducer(initial, action);
		expect(result).toMatchSnapshot();
	});
	it('should handle REPLACE_SELECTED_PHOTOS', () => {
		const action = {
			type: 'REPLACE_SELECTED_PHOTOS',
			payload: [photo1.image.uri],
		};
		const result = devicePhotosReducer(initialState, action);
		expect(result).toMatchSnapshot();
	});
	it('should handle TOGGLE_FETCHING_PHOTOS', () => {
		const action = {
			type: 'TOGGLE_FETCHING_PHOTOS',
			payload: '/photo/uri',
		};
		const result = devicePhotosReducer(initialState, action);
		expect(result).toMatchSnapshot();
	});
	it('should handle TOGGLE_PHOTO_SELECTED', () => {
		const action = {
			type: 'TOGGLE_PHOTO_SELECTED',
			payload: '/photo/uri',
		};
		const initial = {
			...initialState,
			photos: [photo1, photo2, photo3],
		};
		const result = devicePhotosReducer(initial, action);
		expect(result).toMatchSnapshot();
	});
});
