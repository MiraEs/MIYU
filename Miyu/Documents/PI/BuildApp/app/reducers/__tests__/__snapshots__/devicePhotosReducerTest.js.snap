// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`devicePhotosReducer should handle ADD_PHOTO 1`] = `
Object {
  "cursor": "",
  "error": "",
  "hasNextPage": false,
  "isFetchingNextPage": false,
  "photos": Array [
    Object {
      "image": Object {
        "uri": "/test/uri",
      },
      "isSelected": true,
    },
  ],
  "selectedPhotos": Array [
    Object {
      "image": Object {
        "uri": "/test/uri",
      },
      "isSelected": true,
    },
  ],
}
`;

exports[`devicePhotosReducer should handle FILTER_SELECTED_PHOTOS 1`] = `
Object {
  "cursor": "",
  "data": Array [
    Object {
      "selected": true,
      "uri": "/selected/photo",
    },
  ],
  "error": "",
  "hasNextPage": false,
  "isFetchingNextPage": false,
  "photos": Array [],
  "selectedPhotos": Array [],
}
`;

exports[`devicePhotosReducer should handle LOAD_DEVICE_PHOTOS_FAIL 1`] = `
Object {
  "cursor": "",
  "error": [Error: load device photos test error],
  "hasNextPage": false,
  "isFetchingNextPage": false,
  "photos": Array [],
  "selectedPhotos": Array [],
}
`;

exports[`devicePhotosReducer should handle LOAD_DEVICE_PHOTOS_SUCCESS 1`] = `
Object {
  "cursor": "u-u-i-d",
  "error": "",
  "hasNextPage": true,
  "isFetchingNextPage": false,
  "photos": Array [
    Object {
      "image": Object {
        "uri": "/uri/1",
      },
      "isSelected": false,
    },
  ],
  "selectedPhotos": Array [],
}
`;

exports[`devicePhotosReducer should handle LOAD_DEVICE_PHOTOS_SUCCESS with selectedPhotos and one page 1`] = `
Object {
  "cursor": "",
  "error": "",
  "hasNextPage": false,
  "isFetchingNextPage": false,
  "photos": Array [
    Object {
      "image": Object {
        "uri": "/photo/uri",
      },
      "isSelected": true,
    },
  ],
  "selectedPhotos": Array [
    Object {
      "image": Object {
        "uri": "/photo/uri",
      },
      "isSelected": true,
    },
  ],
}
`;

exports[`devicePhotosReducer should handle REPLACE_SELECTED_PHOTOS 1`] = `
Object {
  "cursor": "",
  "error": "",
  "hasNextPage": false,
  "isFetchingNextPage": false,
  "photos": Array [],
  "selectedPhotos": Array [
    "/photo/uri",
  ],
}
`;

exports[`devicePhotosReducer should handle TOGGLE_FETCHING_PHOTOS 1`] = `
Object {
  "cursor": "",
  "error": "",
  "hasNextPage": false,
  "isFetchingNextPage": true,
  "photos": Array [],
  "selectedPhotos": Array [],
}
`;

exports[`devicePhotosReducer should handle TOGGLE_PHOTO_SELECTED 1`] = `
Object {
  "cursor": "",
  "error": "",
  "hasNextPage": false,
  "isFetchingNextPage": false,
  "photos": Array [
    Object {
      "image": Object {
        "uri": "/photo/uri",
      },
      "isSelected": false,
    },
    Object {
      "image": Object {
        "uri": "/other/photo/uri",
      },
      "isSelected": false,
    },
    Object {
      "image": Object {
        "uri": "/yet/another/photo/uri",
      },
      "isSelected": true,
    },
  ],
  "selectedPhotos": Array [
    "/yet/another/photo/uri",
  ],
}
`;
